src/main/java/com/zendesk/maxwell/producer/MaxwellKinesisProducer.java

```diff
@@ -32,82 +31,169 @@
 	private final String json;
 	private final String key;
 
-	public KinesisCallback(AbstractAsyncProducer.CallbackCompleter cc, BinlogPosition position, String key, String json) {
+	private AtomicInteger successRecords;
+	private AtomicInteger errorRecords;
+	private final KinesisProducer kinesisProducer;
+	private final String kinesisStream;
+	private int attempts = 0;
+	private final int maxAttempts;
+
+	public KinesisCallback(AbstractAsyncProducer.CallbackCompleter cc, BinlogPosition position, String key, String json, AtomicInteger successRecords, AtomicInteger errorRecords, KinesisProducer kinesisProducer, String kinesisStream, int attempts, int maxAttempts) {
 		this.cc = cc;
 		this.position = position;
 		this.key = key;
 		this.json = json;
+		this.successRecords = successRecords;
+		this.errorRecords = errorRecords;
+		this.kinesisProducer = kinesisProducer;
+		this.kinesisStream = kinesisStream;
+		this.attempts = attempts;
+		this.maxAttempts = maxAttempts;
 	}
 
 	@Override
 	public void onFailure(Throwable t) {
 		logger.error(t.getClass().getSimpleName() + " @ " + position + " -- " + key);
 		logger.error(t.getLocalizedMessage());
 
-		if(t instanceof UserRecordFailedException) {
-			Attempt last = Iterables.getLast(((UserRecordFailedException) t).getResult().getAttempts());
-			logger.error(String.format("Record failed to put - %s : %s", last.getErrorCode(), last.getErrorMessage()));
-		}
+		if(attempts>=maxAttempts) {
```

Comment:
@alejobs this is a good point since this will mean that the number of retries for a record wont be maxAttempts but instead kplRetries * maxAttempts

I don't have much opinion on this topic, but since the KPL does include retry logic, I personally wouldn't include it unless I encountered some kind issue and figured that the retry knobs that KPL exposes aren't enough to fix it.